src/trace.c

```diff
@@ -1040,6 +1046,38 @@ trace_add_path (const void *parent,
 	if (! path_hash)
 		path_hash = NIH_MUST (nih_hash_string_new (NULL, 2500));
 
+	/* Start checking the file type to see if it can be
+	 * resolved down to some ordinary file.
+	 */
+	if (lstat (pathname, &statbuf) < 0) {
+		return 0;
+	}
+
+	if (! S_ISREG (statbuf.st_mode)) {
+		/* Not an ordinary file. check if it is a symlink,
+		 * and resolve to a file that symlink points to.
+		 */
+		if (S_ISLNK (statbuf.st_mode)) {
+			memset (resolved_pathname, 0, sizeof (resolved_pathname));
+			if (! realpath (pathname, resolved_pathname)) {
+				return 0;
+			}
+
+			pathname = resolved_pathname;
+
+			/* Make sure that the file symlink points to
+			 * is also an ordinary file.
+			 */
+			if ((stat (pathname, &statbuf) < 0)
+			    || (! S_ISREG (statbuf.st_mode))) {
```

Comment:
I wonder if we should check if this is also a symbolic link. It's quite common to have that:
```
$ ls -l /usr/lib |grep -- 'libvpx'
lrwxrwxrwx   1 root root       15 Aug 14  2012 libvpx.so.0 -> libvpx.so.0.9.1
lrwxrwxrwx   1 root root       15 Aug 14  2012 libvpx.so.0.9 -> libvpx.so.0.9.1
-rw-r--r--   1 root root   537632 Nov 12  2010 libvpx.so.0.9.1
```

Reply:
> I wonder if we should check if this is also a symbolic link.

`realpath` resolves nested symlink in one go into a targeted file.  it's man page description says:
```
The resulting path will have no symbolic link,
       /./ or /../ components.
```

> Is it a requirement that you memset the buffer to realpath to zero?

It is one of the bad habit I have, along with resetting everything with ` = {0}`. `realpath` puts null-terminated string for us, so this should not be a requirement. (It's probably unnecessary to put `+1` in the buffer definition either?)

```
The resulting pathname is stored as a null-terminated string, up to a
maximum of PATH_MAX bytes, in the buffer pointed to by
resolved_path.
```

Reply:
Testing locally.
```sh
$ ls -go
total 8
-rw-r----- 1   7 Nov 22 10:08 file
-rw-r----- 1 117 Nov 22 10:01 realpath_test.c
lrwxrwxrwx 1   8 Nov 22 10:08 symlink1 -> symlink2
lrwxrwxrwx 1   4 Nov 22 10:08 symlink2 -> file
```

```
$ cat realpath_test.c
#include <stdlib.h>
#include <stdio.h>

int main(void)
{
        printf("%s\n", realpath("./symlink1", NULL));
        return 0;
}
```

```
$ gcc ./realpath_test.c && ./a.out
/tmp/symlink-check/file
```

Reply:
> memset

Done.